### DISCUSSION TITLE
Benchmarks of rdf4cpp compared to other RDF libraries

### DISCUSSION THREAD

**Discussion by KonradHoeffner** [TortugaAttack/rdf4cpp#213] [Show and tell]
## Query performance

|    | Tool            |   Memory in MB |   Loading time in ms |   Query time in ms |
|---:|:----------------|----------:|---------:|----------:|
|HDT|
|  0 | hdt_cpp         |       **112** |     2050 |       363 |
|  1 | hdt_rs          |       246 |      961 |       340 |
|  2 | sophia_hdt      |       247 |      937 |       340 |
|  3 | hdt_java        |       636 |     3140 |       **260** |
|Non-HDT|
|  4 | sophia_lg       |       **833** |    12220 |        89 |
|  5 | sophia          |      1369 |    16867 |        **22** |
|  6 | **rdf4cpp**         |      1260 |    33436 |      1006 |
|  7 | jena (java)     |      7060 |    33952 |       105 |
|  8 | n3js (js)       |     12406 |    71094 |       546 |
|  9 | rdflib (python) |     14428 |   155455 |       876 |

Time and Memory to load and query 10M triples with a (*,p,o) pattern on an Intel i9-12900k.

## Non-HDT Tools
![Screenshot from 2023-09-29 13-19-04](https://github.com/rdf4cpp/rdf4cpp/assets/839577/4b401e3a-7b47-4f1e-b57b-13529752a840)

## All Tools
![Screenshot from 2023-09-29 13-16-34](https://github.com/rdf4cpp/rdf4cpp/assets/839577/70191635-585a-42f4-91d9-378d7bd2ab37)

## HDT Tools
![Screenshot from 2023-09-29 13-19-15](https://github.com/rdf4cpp/rdf4cpp/assets/839577/15f010f4-f19f-4c01-9e9d-a813a0cee8fd)

See <https://github.com/KonradHoeffner/hdt_benchmark/blob/master/benchmark_results.ipynb> for details.
Thanks to @bigerl for the integration.
Feel free to [create an issue](https://github.com/KonradHoeffner/hdt_benchmark/issues) or a [pull request](https://github.com/KonradHoeffner/hdt_benchmark/pulls) if you spot an error or area for improvement in the benchmark code.

**Comment by bigerl** [comment]
Thank you for sharing @KonradHoeffner .
Very interesting to see. With rdf4cpp we focused so far mainly on correct support of RDF literals with all their type hierarchies, casting, functions and operations. Querying is really not a focus until now, a graph is a simple set of and scanned linearly to resolve triple patterns. 

For parsing we rely so far on [serd](https://github.com/drobilla/serd) for parsing. Seeing how Sophia performs I believe we should probably investigate how we could improve the parsing performance. Not sure if the problem is serd or what we do afterwards with it's output.

**Reply by KonradHoeffner** [reply]
Linear scanning sounds like binary search could be an option to improve at least some triple patterns.
As for the parsing performance bottlenecks, I tried to profile it with perf but the call graph in Firefox profiler only shows a single level.
Trying it again with a debug build with and with `perf record -g --call-graph dwarf -- myapp`.
That is much slower with 173s instead of ~33s for the release build but hopefully the proportions are similar.
But --call-graph dwarf also is missing too much, only --call-graph lbr provides sufficient detail.

## Release, 10M triples
![Screenshot from 2023-09-29 14-31-25](https://github.com/rdf4cpp/rdf4cpp/assets/839577/b6a699d6-9a0d-42d6-8aeb-c9b029402563)

## Debug, 10M triples
![Screenshot from 2023-09-29 14-32-33](https://github.com/rdf4cpp/rdf4cpp/assets/839577/f42b299f-d168-4003-bde3-feef4dea5a5d)

## Debug with call graph lbr, 1M triples
![Screenshot from 2023-09-29 14-36-46](https://github.com/rdf4cpp/rdf4cpp/assets/839577/067434d0-4d6d-4855-be40-ed973123f529)
![Screenshot from 2023-09-29 14-37-44](https://github.com/rdf4cpp/rdf4cpp/assets/839577/2f6ef9e6-aa83-454f-a80e-27311f393b04)

## Debug, LBR, 10M triples

The profile gets to 3.9GB and Firefox Profiler can't read it, maybe there is a maximum size?

## Release, LBR, 10M Triples
![Screenshot from 2023-09-29 14-43-35](https://github.com/rdf4cpp/rdf4cpp/assets/839577/5c97d707-306e-4bc7-99da-b350227727d7)
![Screenshot from 2023-09-29 14-44-12](https://github.com/rdf4cpp/rdf4cpp/assets/839577/5e0a6bfa-2d8e-42d5-8a8f-74394d209520)

**Reply by bigerl** [reply]
Hmm, my first take-away is: we should not use a `std::map` because correctly ordering rdf nodes is just too expensive. 

-> https://github.com/rdf4cpp/rdf4cpp/issues/215
